source/slang/glsl.meta.slang

```diff
@@ -2825,21 +2825,1988 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
+            return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformSMin $$T result Subgroup ExclusiveScan $value};
+        else if (__isUnsignedInt<T>())
+            return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformUMin $$T result Subgroup ExclusiveScan $value};
+        else return value;
+    case hlsl:
+    case cuda:
+    case cpp:
+        // TODO: implementation
+        return value;
+    }
+    return value;
+}
+
+__generic<T : __BuiltinArithmeticType>
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_arithmetic)
+spirv_caps_shader_subgroup(1.3)
+[ForceInline] public T subgroupExclusiveMax(T value)
+{
+    shader_subgroup_preamble<T>();
+    __target_switch
+    {
+    case glsl:
+        __target_intrinsic "subgroupExclusiveMax($0)";
+    case spirv:
+        if (__isFloat<T>())
+            return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformFMax $$T result Subgroup ExclusiveScan $value};
+        else if (__isSignedInt<T>())
+            return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformSMax $$T result Subgroup ExclusiveScan $value};
+        else if (__isUnsignedInt<T>())
+            return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformUMax $$T result Subgroup ExclusiveScan $value};
+        else return value;
+    case hlsl:
+    case cuda:
+    case cpp:
+        // TODO: implementation
+        return value;
+    }
+    return value;
+}
+
+__generic<T : __BuiltinLogicalType>
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_arithmetic)
+spirv_caps_shader_subgroup(1.3)
+[ForceInline] public T subgroupExclusiveAnd(T value)
+{
+    shader_subgroup_preamble<T>();
+    __target_switch
+    {
+    case glsl: __intrinsic_asm "subgroupExclusiveAnd($0)";
+    case spirv:
+        if (__isBool<T>()) return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformLogicalAnd $$T result Subgroup ExclusiveScan $value};
+        else return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformBitwiseAnd $$T result Subgroup ExclusiveScan $value};
+    case hlsl:
+    case cuda:
+    case cpp:
+        // TODO: implementation
+        return value;
+    }
+}
+
+__generic<T : __BuiltinLogicalType>
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_arithmetic)
+spirv_caps_shader_subgroup(1.3)
+[ForceInline] public T subgroupExclusiveOr(T value)
+{
+    shader_subgroup_preamble<T>();
+    __target_switch
+    {
+    case glsl: __intrinsic_asm "subgroupExclusiveOr($0)";
+    case spirv:
+        if (__isBool<T>()) return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformLogicalOr $$T result Subgroup ExclusiveScan $value};
+        else return spirv_asm {OpCapability GroupNonUniformArithmetic; OpGroupNonUniformBitwiseOr $$T result Subgroup ExclusiveScan $value};
+    case hlsl:
+    case cuda:
+    case cpp:
+        // TODO: implementation
+        return value;
+    }
+}
```

Comment:
Is seems like there is WaveMaskPrefixBitOr in hlsl.meta.slang.
Can we just call that?

Reply:
we cannot since as per spec:
```
genIType subgroupExclusiveOr(genIType value);
genUType subgroupExclusiveOr(genUType value);
genBType subgroupExclusiveOr(genBType value);
```
where bool types do logical operations rather than bitwise; this is different from HLSL which does bitwise operations (bool(0x10) | bool(0x10) will work differently across HLSL & GLSL)

there is another issue with hlsl.meta using __BuiltinArithmeticType, although I believe hlsl.meta should be using __BuiltinLogicalType (waveallbitor is defined with __int_type__, but this was not looked into)